### ISSUE TITLE
OpenCL data type inconsistencies

### ISSUE THREAD

**Issue by MichaelClerx** [myokit/myokit#902]
Causing warnings, but also wrong output for e.g. `myokit opencl`:

```

C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(698): warning C4244: 'function': conversion from 'cl_ulong' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(705): warning C4244: 'function': conversion from 'cl_ulong' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(712): warning C4244: 'function': conversion from 'cl_ulong' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(726): warning C4267: 'function': conversion from 'size_t' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(742): warning C4267: 'function': conversion from 'size_t' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(751): warning C4267: 'function': conversion from 'size_t' to 'long', possible loss of data
   Creating library build\temp.win-amd64-cpython-39\Release\Users\C8B69~1.MEE\AppData\Local\Temp\tmpus8x_zkymyokit\myokit_opencl_info_1_5578598959800797075.cp39-win_amd64.lib and object build\temp.win-amd64-cpython-39\Release\Users\C8B69~1.MEE\AppData\Local\Temp\tmpus8x_zkymyokit\myokit_opencl_info_1_5578598959800797075.cp39-win_amd64.exp
Generating code
Finished generating code
Platform 0
 Name       : NVIDIA CUDA
 Vendor     : NVIDIA Corporation
 Version    : OpenCL 3.0 CUDA 11.8.88
 Profile    : FULL_PROFILE
 Extensions : cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_fp64 cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_icd cl_khr_gl_sharing cl_nv_compiler_options cl_nv_device_attribute_query cl_nv_pragma_unroll cl_nv_d3d10_sharing cl_khr_d3d10_sharing cl_nv_d3d11_sharing cl_nv_copy_opts cl_khr_gl_event cl_nv_create_buffer cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_nv_kernel_attribute cl_khr_device_uuid cl_khr_pci_bus_info cl_khr_external_semaphore cl_khr_external_memory cl_khr_external_semaphore_win32 cl_khr_external_memory_win32
 Devices:
  Device 0
   Name            : Quadro P2000
   Vendor          : NVIDIA Corporation
   Version         : OpenCL 3.0 CUDA
   Driver          : 522.06
   Clock speed     : 1607 MHz
   Global memory   : -196608 B
   Local memory    : 48 KB
   Constant memory : 64 KB
   Max param size  : 4352 bytes
   Max work groups : 1024
   Max work items  : [1024, 1024, 64]
Platform 1
 Name       : Intel(R) OpenCL HD Graphics
 Vendor     : Intel(R) Corporation
 Version    : OpenCL 3.0
 Profile    : FULL_PROFILE
 Extensions : cl_khr_byte_addressable_store cl_khr_fp16 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_icd cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_intel_command_queue_families cl_intel_subgroups cl_intel_required_subgroup_size cl_intel_subgroups_short cl_khr_spir cl_intel_accelerator cl_intel_driver_diagnostics cl_khr_priority_hints cl_khr_throttle_hints cl_khr_create_command_queue cl_intel_subgroups_char cl_intel_subgroups_long cl_khr_il_program cl_intel_mem_force_host_memory cl_khr_subgroup_extended_types cl_khr_subgroup_non_uniform_vote cl_khr_subgroup_ballot cl_khr_subgroup_non_uniform_arithmetic cl_khr_subgroup_shuffle cl_khr_subgroup_shuffle_relative cl_khr_subgroup_clustered_reduce cl_intel_device_attribute_query cl_khr_suggested_local_work_size cl_intel_split_work_group_barrier cl_khr_fp64 cl_khr_subgroups cl_intel_spirv_device_side_avc_motion_estimation cl_intel_spirv_media_block_io cl_intel_spirv_subgroups cl_khr_spirv_no_integer_wrap_decoration cl_intel_unified_shared_memory cl_khr_mipmap_image cl_khr_mipmap_image_writes cl_intel_planar_yuv cl_intel_packed_yuv cl_intel_motion_estimation cl_intel_device_side_avc_motion_estimation cl_intel_advanced_motion_estimation cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_3d_image_writes cl_intel_media_block_io cl_khr_gl_sharing cl_khr_gl_depth_images cl_khr_gl_event cl_khr_gl_msaa_sharing cl_intel_dx9_media_sharing cl_khr_dx9_media_sharing cl_khr_d3d10_sharing cl_khr_d3d11_sharing cl_intel_d3d11_nv12_media_sharing cl_intel_sharing_format_query cl_khr_pci_bus_info cl_intel_simultaneous_sharing
 Devices:
  Device 0
   Name            : Intel(R) UHD Graphics 630
   Vendor          : Intel(R) Corporation
   Version         : OpenCL 3.0 NEO
   Driver          : 31.0.101.2111
   Clock speed     : 1150 MHz
   Global memory   : -1797140480 B
   Local memory    : 64 KB
   Constant memory : -898570240 B
   Max param size  : 2048 bytes
   Max work groups : 256
   Max work items  : [256, 256, 256]​
```

Note the negative global & constant memory sizes

**Comment by MichaelClerx** [comment]
I suspect this is also the cause of this error:

```
running build_ext
building 'myokit_sim_opencl_1_7309369576212041806' extension
creating build
creating build\temp.win-amd64-cpython-39
creating build\temp.win-amd64-cpython-39\Release
creating build\temp.win-amd64-cpython-39\Release\Users
creating build\temp.win-amd64-cpython-39\Release\Users\username
creating build\temp.win-amd64-cpython-39\Release\Users\username\AppData
creating build\temp.win-amd64-cpython-39\Release\Users\username\AppData\Local
creating build\temp.win-amd64-cpython-39\Release\Users\username\AppData\Local\Temp
creating build\temp.win-amd64-cpython-39\Release\Users\username\AppData\Local\Temp\tmpecheldo6myokit
"C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8\include" -IC:\Users\username\miniconda3\lib\site-packages\myokit\_sim -IC:\Users\username\miniconda3\lib\site-packages\myokit\_sim -IC:\Users\username\miniconda3\include -IC:\Users\username\miniconda3\Include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /TcC:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c /Fobuild\temp.win-amd64-cpython-39\Release\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.obj -Wall
source.c
...
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(410): warning C4456: declaration of 'i' hides previous local declaration
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(288): note: see declaration of 'i'
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(432): warning C4456: declaration of 'device_ids' hides previous local declaration
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(284): note: see declaration of 'device_ids'
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(433): warning C4456: declaration of 'n_devices' hides previous local declaration
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(285): note: see declaration of 'n_devices'
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(461): warning C4456: declaration of 'device_ids' hides previous local declaration
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(284): note: see declaration of 'device_ids'
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(462): warning C4456: declaration of 'n_devices' hides previous local declaration
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(285): note: see declaration of 'n_devices'
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(463): warning C4456: declaration of 'j' hides previous local declaration
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(288): note: see declaration of 'j'
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(698): warning C4242: 'function': conversion from 'cl_ulong' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(705): warning C4242: 'function': conversion from 'cl_ulong' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(712): warning C4242: 'function': conversion from 'cl_ulong' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(726): warning C4267: 'function': conversion from 'size_t' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(742): warning C4267: 'function': conversion from 'size_t' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(751): warning C4267: 'function': conversion from 'size_t' to 'long', possible loss of data
C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\mcl.h(795): warning C4255: 'mcl_info': no function prototype given: converting '()' to '(void)'
C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c(184): warning C4255: 'sim_clean': no function prototype given: converting '()' to '(void)'
C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c(260): warning C4100: 'args': unreferenced formal parameter
C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c(260): warning C4100: 'self': unreferenced formal parameter
C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c(427): warning C4242: '=': conversion from 'Py_ssize_t' to 'unsigned long', possible loss of data
C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c(436): warning C4242: '=': conversion from 'Py_ssize_t' to 'unsigned long', possible loss of data
C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c(601): warning C4242: '=': conversion from 'Py_ssize_t' to 'unsigned long', possible loss of data
C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c(885): warning C4242: '=': conversion from 'Py_ssize_t' to 'unsigned long', possible loss of data
C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c(276): warning C4100: 'self': unreferenced formal parameter
C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c(1077): warning C4100: 'args': unreferenced formal parameter
C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.c(1077): warning C4100: 'self': unreferenced formal parameter
creating C:\Users\username\AppData\Local\Temp\tmpecheldo6myokit\build\build\lib.win-amd64-cpython-39
"C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\bin\HostX86\x64\link.exe" /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO "/LIBPATH:C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8\lib\x64" /LIBPATH:C:\Users\username\miniconda3\libs /LIBPATH:C:\Users\username\miniconda3 /LIBPATH:C:\Users\username\miniconda3\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\um\x64" OpenCL.lib /EXPORT:PyInit_myokit_sim_opencl_1_7309369576212041806 build\temp.win-amd64-cpython-39\Release\Users\username\AppData\Local\Temp\tmpecheldo6myokit\source.obj /OUT:build\lib.win-amd64-cpython-39\myokit_sim_opencl_1_7309369576212041806.cp39-win_amd64.pyd /IMPLIB:build\temp.win-amd64-cpython-39\Release\Users\username\AppData\Local\Temp\tmpecheldo6myokit\myokit_sim_opencl_1_7309369576212041806.cp39-win_amd64.lib
   Creating library build\temp.win-amd64-cpython-39\Release\Users\username\AppData\Local\Temp\tmpecheldo6myokit\myokit_sim_opencl_1_7309369576212041806.cp39-win_amd64.lib and object build\temp.win-amd64-cpython-39\Release\Users\username\AppData\Local\Temp\tmpecheldo6myokit\myokit_sim_opencl_1_7309369576212041806.cp39-win_amd64.exp
Generating code
...
Finished generating code
copying build\lib.win-amd64-cpython-39\myokit_sim_opencl_1_7309369576212041806.cp39-win_amd64.pyd ->
Traceback (most recent call last):
  File "C:\Users\username\miniconda3\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\username\miniconda3\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\username\miniconda3\lib\site-packages\myokit\__main__.py", line 2207, in <module>
    main()
  File "C:\Users\username\miniconda3\lib\site-packages\myokit\__main__.py", line 85, in main
    func(**args)
  File "C:\Users\username\miniconda3\lib\site-packages\myokit\__main__.py", line 1032, in run
    myokit.run(model, protocol, script)
  File "C:\Users\username\miniconda3\lib\site-packages\myokit\_aux.py", line 523, in run
    r.run()
  File "C:\Users\username\miniconda3\lib\site-packages\myokit\_aux.py", line 514, in run
    myokit._exec(self.script, environment)
  File "C:\Users\username\miniconda3\lib\site-packages\myokit\_exec_new.py", line 15, in _exec
    exec(script, globals, locals)
  File "<string>", line 21, in <module>
  File "C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\openclsim.py", line 1002, in run
    r = self._run(duration, log, log_interval, report_nan, progress, msg)
  File "C:\Users\username\miniconda3\lib\site-packages\myokit\_sim\openclsim.py", line 1094, in _run
    self._sim.sim_init(
Exception: OpenCL error: CL_INVALID_ARG_SIZE
```

**Comment by MichaelClerx** [comment]
https://www.intel.com/content/www/us/en/support/programmable/articles/000086479.html

**Comment by MichaelClerx** [comment]
Nope. Two unrelated issues! Both solved in #903
